File: apps/backend/internal/model/analytic.go/expense_analytics.go
package analytic

import (
	"github.com/xanity-07/spndex/internal/enums"
	"github.com/xanity-07/spndex/internal/model/expense"
)

type CategoryTotals struct {
	Category enums.ExpenseCategory `json:"category" db:"category"`
	CategoryStats
	Percentage float64 `json:"percentage" db:"-"`
}

type CategoryStats struct {
	TotalCents int64 `json:"-" db:"total"`
	Count      int64 `json:"count" db:"count"`
}

type MonthlyTotals struct {
	Month      string `json:"month" db:"month"`
	TotalCents int64  `json:"totalCents" db:"total"`
	Count      int64  `json:"count" db:"count"`
}

type DashboardStats struct {
	HighestExpense            expense.Expense `json:"highestExpense"`
	LowestExpense             expense.Expense `json:"lowestExpense"`
	TotalExpensesCents        int64           `json:"totalExpenses"`
	ExpenseCount              int             `json:"expenseCount"`
	AverageExpenseAmountCents int64           `json:"averageExpenseAmountCents"`
	CurrentMonthTotalCents    int64           `json:"currentMonthTotalCents"`
	LastMonthTotalCents       int64           `json:"lastMonthTotalCents"`
	MonthlyNetChangeCents     int64           `json:"monthlyNetChangeCents"`
}

File: packages/zod/src/expenseAnalytics.ts
import * as z from 'zod';

import { ZExpense, ZExpenseCategory } from './expense.js';

export const ZCategoryTotals = z.object({
    category: ZExpenseCategory,

    totalCents: z
        .number()
        .int()
        .nonnegative()
        .meta({
            description:
                'Total expense amount for the category in cents. For example, 45000 represents 450.00 units of the applicable currency.',
            examples: [45000],
        }),

    count: z
        .number()
        .int()
        .nonnegative()
        .meta({
            description: 'Number of expenses recorded in the category.',
            examples: [23],
        }),

    percentage: z
        .number()
        .min(0)
        .max(100)
        .meta({
            description: 'Percentage of total expenses represented by the category.',
            examples: [35.42],
        }),
});

export const ZMonthlyTotals = z.object({
    month: z
        .string()
        .regex(/^\d{4}-(0[1-9]|1[0-2])$/)
        .meta({
            description: 'Month represented in YYYY-MM format.',
            examples: ['2026-08'],
        }),

    totalCents: z
        .number()
        .int()
        .nonnegative()
        .meta({
            description:
                'Total expense amount for the month in cents. For example, 125000 represents 1250.00 units of the applicable currency.',
            examples: [125000],
        }),

    count: z
        .number()
        .int()
        .nonnegative()
        .meta({
            description: 'Number of expenses recorded during the month.',
            examples: [42],
        }),
});

export const ZDashboardStats = z.object({
    highestExpense: ZExpense.meta({
        description: 'The highest-value expense recorded by the user.',
    }),

    lowestExpense: ZExpense.meta({
        description: 'The lowest-value expense recorded by the user.',
    }),

    totalExpensesCents: z
        .number()
        .int()
        .nonnegative()
        .meta({
            description:
                'Total amount of all expenses in cents. For example, 125000 represents 1250.00 units of the applicable currency.',
            examples: [125000],
        }),

    expenseCount: z
        .number()
        .int()
        .nonnegative()
        .meta({
            description: 'Total number of expenses recorded by the user.',
            examples: [42],
        }),

    averageExpenseAmountCents: z
        .number()
        .int()
        .nonnegative()
        .meta({
            description:
                'Average expense amount in cents. For example, 2976 represents 29.76 units of the applicable currency.',
            examples: [2976],
        }),

    currentMonthTotalCents: z
        .number()
        .int()
        .nonnegative()
        .meta({
            description:
                'Total expense amount for the current month in cents. For example, 85000 represents 850.00 units of the applicable currency.',
            examples: [85000],
        }),

    lastMonthTotalCents: z
        .number()
        .int()
        .nonnegative()
        .meta({
            description:
                'Total expense amount for the previous month in cents. For example, 72000 represents 720.00 units of the applicable currency.',
            examples: [72000],
        }),

    monthlyNetChangeCents: z
        .number()
        .int()
        .meta({
            description:
                'Net change in expense spending between the current and previous month in cents. Positive values indicate an increase, while negative values indicate a decrease.',
            examples: [13000],
        }),
});
